utilities/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+            // Report 1-based line number, rather than 0-based
+            f(i+1, &items);
+        });
+}
+
+pub struct RecordDeltaEntry {
+    pub diff_abs: f64,
+    pub diff_rel: f64,
+    pub sign_change: bool,
+    pub line_abs: usize,
+    pub line_rel: usize,
+}
+
+impl RecordDeltaEntry {
+    pub fn new() -> Self {
+        RecordDeltaEntry {
+            diff_abs: 0.0,
+            diff_rel: 0.0,
+            sign_change: false,
+            line_abs: 0,
+            line_rel: 0,
+        }
+    }
+}
+
+impl Copy for RecordDeltaEntry {}
+
+impl Clone for RecordDeltaEntry {
+        fn clone(&self) -> Self {
+            RecordDeltaEntry {
+                diff_abs: self.diff_abs,
+                diff_rel: self.diff_rel,
+                sign_change: self.sign_change,
+                line_abs: self.line_rel,
+                line_rel: self.line_rel,
+            }
+        }
+}
+
+pub fn record_delta(x: f64, y: f64, line_num: usize, entry: &mut RecordDeltaEntry) {
+    let delta2 = calc_delta(x, y);
+    if (f64::is_nan(delta2.0) && !f64::is_nan(entry.diff_abs)) || delta2.0 > entry.diff_abs {
+        entry.diff_abs = delta2.0;
+        entry.line_abs = line_num;
+    }
+    if (f64::is_nan(delta2.1) && !f64::is_nan(entry.diff_rel)) || delta2.1 > entry.diff_rel {
+        entry.diff_rel = delta2.1;
+        entry.line_rel = line_num;
+    }
+    if x.is_sign_negative() != y.is_sign_negative() {
+        entry.sign_change = true;
+    }
+}
+
+// Check whether x and y are within delta of each other, considering both absolute and relative difference.
+// Consider them close enough if absolute difference is within delta inclusive,
+// or if relative difference is within delta exclusive.
+// nan values are considered equal to each other, and outside delta from everything else.
+#[macro_export]
+macro_rules! assert_delta {
+    ($x:expr, $y:expr, $delta:expr, $allow_sign_change:expr, $msg:expr, $line_num:expr) => {
+            // todo: use the same formatting behavior as "assert!", instead of line_num hack
+            // todo: consider using separate deltas for relative and absolute
+            let delta2 = utilities::calc_delta($x, $y);
+            assert!(
+                    delta2.0 <= $delta || delta2.1 < $delta,
+                    "assert_delta failed line {}, {}: {:e} vs {:e} diff {:e} rel {:e} outside inclusive {:e}", $line_num, $msg, $x, $y, delta2.0, delta2.1, $delta
+            );
+            // For the sign change check, allow (NAN vs NAN), but not (0.0 vs -0.0) or (NAN vs -NAN).
+            assert!(
+                $allow_sign_change || $x.is_sign_negative() == $y.is_sign_negative(),
+                "assert_delta failed line {}, {}: {:e} vs {:e} sign difference disallowed", $line_num, $msg, $x, $y
+            );
+    }
+}
+
+
+
+// fn unescape_all(s: &String) -> String {
```

Comment:
What is this for?

Reply:
That's for later operations that use string arguments that might include spaces. I meant to stash that somewhere before my last commit, and remove it from committed files, but apparently forgot to do that.